From 3f8f0bb7ccc4aaef4d56fbe9dbd81ee89c294520 Mon Sep 17 00:00:00 2001
From: Steve Korshakov <korshakov.stepan@gmail.com>
Date: Mon, 3 Apr 2023 19:14:27 +0100
Subject: [PATCH] Update CMakeLists.txt

---
 emulator/CMakeLists.txt | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/emulator/CMakeLists.txt b/emulator/CMakeLists.txt
index 2fb27519d..67d994ff7 100644
--- a/emulator/CMakeLists.txt
+++ b/emulator/CMakeLists.txt
@@ -51,5 +51,8 @@ if (USE_EMSCRIPTEN)
   target_link_options(emulator-emscripten PRIVATE -sENVIRONMENT=web)
   target_link_options(emulator-emscripten PRIVATE -sFILESYSTEM=0)
   target_link_options(emulator-emscripten PRIVATE -fexceptions)
+if (USE_EMSCRIPTEN_NO_WASM)
+  target_link_options(emulator-emscripten PRIVATE -sWASM=0)
+endif()
   target_compile_options(emulator-emscripten PRIVATE -fexceptions)
 endif()
